rp2040-hal/src/gpio/pin/non_blocking.rs

```diff
@@ -0,0 +1,200 @@
+            || pin.interrupt_status(Interrupt::EdgeLow)
+            || pin.interrupt_status(Interrupt::EdgeHigh)
+        {
+            pin.set_interrupt_enabled(Interrupt::LevelLow, false);
+            pin.set_interrupt_enabled(Interrupt::LevelHigh, false);
+            pin.set_interrupt_enabled(Interrupt::EdgeLow, false);
+            pin.set_interrupt_enabled(Interrupt::EdgeHigh, false);
+
+            pin.clear_interrupt(Interrupt::LevelLow);
+            pin.clear_interrupt(Interrupt::LevelHigh);
+            pin.clear_interrupt(Interrupt::EdgeLow);
+            pin.clear_interrupt(Interrupt::EdgeHigh);
+
+            Self::waker().wake();
+        }
+    }
+}
+
+impl<I, P, S> Wakeable for Pin<I, FunctionSio<S>, P>
+where
+    I: PinId + Wakeable,
+    P: PullType,
+    S: SioConfig,
+{
+    fn waker() -> &'static IrqWaker {
+        I::waker()
+    }
+}
+
+impl<I, P, S> Pin<I, FunctionSio<S>, P>
+where
+    I: PinId + Wakeable,
+    P: PullType,
+    S: SioConfig,
+{
+    fn poll_is_high(&mut self) -> Poll<Result<(), Error>> {
+        if self._is_high() {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
+        }
+    }
+
+    fn poll_is_low(&mut self) -> Poll<Result<(), Error>> {
+        if self._is_low() {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
+        }
+    }
+
+    fn disable_level_high_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::LevelHigh, false);
+    }
+
+    fn enable_level_high_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::LevelHigh, true);
+    }
+
+    fn disable_level_low_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::LevelLow, false);
+    }
+
+    fn enable_level_low_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::LevelLow, true);
+    }
+
+    fn disable_rising_edge_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::EdgeHigh, false)
+    }
+
+    fn enable_rising_edge_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::EdgeHigh, true)
+    }
+
+    fn disable_falling_edge_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::EdgeLow, false)
+    }
+
+    fn enable_falling_edge_irq(&mut self) {
+        self.set_interrupt_enabled(Interrupt::EdgeLow, true)
+    }
+
+    async fn _wait_for_high(&mut self) {
+        let _ = CPFn::new(
+            self,
+            Self::poll_is_high,
+            Self::enable_level_high_irq,
+            Self::disable_level_high_irq,
+        )
+        .await;
+    }
+
+    async fn _wait_for_low(&mut self) {
+        let _ = CPFn::new(
+            self,
+            Self::poll_is_low,
+            Self::enable_level_low_irq,
+            Self::disable_level_low_irq,
+        )
+        .await;
+    }
+
+    async fn _wait_for_rising_edge(&mut self) {
+        if self._is_high() {
+            self._wait_for_low().await;
+        }
+
+        let _ = CPFn::new(
+            self,
+            Self::poll_is_high,
```

Comment:
I don't think "wait for rising edge" can use poll_is_high as its poll function. A high level does not imply that an edge has occurred.
Better read the INTR register. (This may require that you don't clear the INTR bit in on_interrupt, but only do it in the poll function. I don't remember the details of the waker interactions.)

Reply:
Initialy before this call there was a code to ensure pin is low. 

I updated the implementation to not clear interrupt inside on_interrupt and use a poll function specific to the edge interrupt